src/storage.rs

```diff
@@ -419,200 +450,324 @@ impl Cache {
-                    let diff = to_len as i64 - from_len as i64;
-                    let count = diff.unsigned_abs();
-                    let raw = if diff < 0 {
-                        format!("-{}", count)
-                    } else {
-                        format!("+{}", count)
-                    };
-                    DiffStat { raw, count }
                 }
-            };
 
-            let rec = DiffRecommendation {
-                from: delta.from.clone(),
-                to: delta.to.clone(),
-                diffstat,
-            };
+                Ok::<_, VetError>(diffstat)
+            })
+            .await?;
+        Ok(diffstat.clone())
+    }
 
-            if let Some(best) = best_rec.as_ref() {
-                if best.diffstat.count > rec.diffstat.count {
-                    best_rec = Some(rec);
-                }
+    /// Run a garbage-collection pass over the cache, removing any files which
+    /// aren't supposed to be there, or which haven't been touched for an
+    /// extended period of time.
+    pub async fn gc(&self, max_package_age: Duration) {
+        if self.root.is_none() {
+            return;
+        }
+
+        let (root_rv, empty_rv, packages_rv) = tokio::join!(
+            self.gc_root(),
+            self.gc_empty(),
+            self.gc_packages(max_package_age)
+        );
+        if let Err(err) = root_rv {
+            error!("gc: performing gc on the cache root failed: {err}");
+        }
+        if let Err(err) = empty_rv {
+            error!("gc: performing gc on the empty package failed: {err}");
+        }
+        if let Err(err) = packages_rv {
+            error!("gc: performing gc on the package cache failed: {err}");
+        }
+    }
+
+    /// Sync version of `gc`
+    pub fn gc_sync(&self, max_package_age: Duration) {
+        tokio::runtime::Handle::current().block_on(self.gc(max_package_age));
+    }
+
+    /// Remove any unrecognized files from the root of the cargo-vet cache
+    /// directory.
+    async fn gc_root(&self) -> Result<(), VetError> {
+        let root = self.root.as_ref().unwrap();
+        let mut root_entries = tokio::fs::read_dir(root).await?;
+        while let Some(entry) = root_entries.next_entry().await? {
+            if !entry
+                .file_name()
+                .to_str()
+                .map_or(false, |name| CACHE_ALLOWED_FILES.contains(&name))
+            {
+                remove_dir_entry(&entry).await?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Remove all files located in the `cargo-vet/empty` directory, as it
+    /// should be empty.
+    async fn gc_empty(&self) -> Result<(), VetError> {
+        let empty = self.root.as_ref().unwrap().join(CACHE_EMPTY_PACKAGE);
+        let mut empty_entries = tokio::fs::read_dir(&empty).await?;
+        while let Some(entry) = empty_entries.next_entry().await? {
+            remove_dir_entry(&entry).await?;
+        }
+        Ok(())
+    }
```

Comment:
actually now that I think about it empty/ should contain a .cargo-ok, but not a problem for this PR.